sycl/include/sycl/ext/oneapi/experimental/complex/detail/marray_math.hpp

```diff
@@ -0,0 +1,191 @@
+    sycl::marray<rtn_type, NumElements> rtn;                                   \
+    for (std::size_t i = 0; i < NumElements; ++i) {                            \
+      rtn[i] = math_func(x, y[i]);                                             \
+    }                                                                          \
+    return rtn;                                                                \
+  }
+
+MARRAY_CPLX_MATH_OP_TWO_PARAM(pow, complex<T>, complex<T>, T);
+MARRAY_CPLX_MATH_OP_TWO_PARAM(pow, complex<T>, complex<T>, complex<T>);
+MARRAY_CPLX_MATH_OP_TWO_PARAM(pow, complex<T>, T, complex<T>);
+
+#undef MARRAY_CPLX_MATH_OP_TWO_PARAM
+
+// Special definition as polar requires default argument
+
+template <typename T, std::size_t NumElements>
+_SYCL_EXT_CPLX_INLINE_VISIBILITY
+    typename std::enable_if_t<is_genfloat<T>::value,
+                              sycl::marray<complex<T>, NumElements>>
+    polar(const sycl::marray<T, NumElements> &rho,
+          const sycl::marray<T, NumElements> &theta) {
+  sycl::marray<complex<T>, NumElements> rtn;
+  for (std::size_t i = 0; i < NumElements; ++i) {
+    rtn[i] = polar(rho[i], theta[i]);
+  }
+  return rtn;
+}
+
+template <typename T, std::size_t NumElements>
+_SYCL_EXT_CPLX_INLINE_VISIBILITY
+    typename std::enable_if_t<is_genfloat<T>::value,
+                              sycl::marray<complex<T>, NumElements>>
+    polar(const sycl::marray<T, NumElements> &rho, const T &theta = 0) {
+  sycl::marray<complex<T>, NumElements> rtn;
+  for (std::size_t i = 0; i < NumElements; ++i) {
+    rtn[i] = polar(rho[i], theta);
+  }
+  return rtn;
+}
+
+template <typename T, std::size_t NumElements>
+_SYCL_EXT_CPLX_INLINE_VISIBILITY
+    typename std::enable_if_t<is_genfloat<T>::value,
+                              sycl::marray<complex<T>, NumElements>>
+    polar(const T &rho, const sycl::marray<T, NumElements> &theta) {
+  sycl::marray<complex<T>, NumElements> rtn;
+  for (std::size_t i = 0; i < NumElements; ++i) {
+    rtn[i] = polar(rho, theta[i]);
+  }
+  return rtn;
+}
+
+// real
+
+template <typename T, std::size_t NumElements>
+_SYCL_EXT_CPLX_INLINE_VISIBILITY
+    typename std::enable_if_t<is_genfloat<T>::value,
+                              sycl::marray<T, NumElements>>
+    real(const marray<complex<T>, NumElements> &z) {
+  sycl::marray<T, NumElements> rtn;
+  for (std::size_t i = 0; i < NumElements; ++i) {
+    rtn[i] = z[i].real();
+  }
+  return rtn;
+}
+
+// imag
+
+template <typename T, std::size_t NumElements>
+_SYCL_EXT_CPLX_INLINE_VISIBILITY
+    typename std::enable_if_t<is_genfloat<T>::value,
+                              sycl::marray<T, NumElements>>
+    imag(const marray<complex<T>, NumElements> &z) {
+  sycl::marray<T, NumElements> rtn;
+  for (std::size_t i = 0; i < NumElements; ++i) {
+    rtn[i] = z[i].imag();
+  }
+  return rtn;
+}
+
+} // namespace experimental
+} // namespace oneapi
+} // namespace ext
+
+} // namespace _V1
+} // namespace sycl
```

Comment:
@aelovikov-intel - With the builtin refactoring, can we reuse some of the new tooling for this?

Reply:
I'd say the `NUM_ARGS` macro parameter approach looks promising, although that's more of an idiom than tooling. Having a `marray_map` helper would remove some code duplication as well.

Reply:
I've updated the builtins and explained the details in my last comment